beacon-chain/rpc/over/handlers_deposit.go

```diff
@@ -0,0 +1,293 @@
+		// If validator is already in the registry and ready to be activated,
+		// expectedActivationEpoch will be set to the epoch when the validator is assigned/expected to be activated.
+		if expectedActivationEpoch < params.BeaconConfig().FarFutureEpoch {
+			data.ExpectedActivationEpoch = uint64(expectedActivationEpoch)
+		}
+	}
+
+	pdes, err := buildPendingDepositEstimations(st, pubkey, primitives.Gwei(activeBalance), !found /* initial */)
+	if err != nil {
+		httputil.WriteError(w, handleWrapError(err, "could not build pending deposit estimations", http.StatusBadRequest))
+		return
+	}
+	data.PendingDeposits = pdes
+
+	httputil.WriteJson(w, &structs.GetDepositEstimationResponse{
+		ExecutionOptimistic: isOptimistic,
+		Finalized:           isFinalized,
+		Data:                data,
+	})
+}
+
+// buildPendingDepositEstimations iterates through the pending deposits and calculates the expected epoch.
+func buildPendingDepositEstimations(st state.BeaconState, pubkey []byte,
+	activeBalance primitives.Gwei, initial bool) ([]*structs.PendingDepositEstimationContainer, error) {
+	pdes := make([]*structs.PendingDepositEstimationContainer, 0)
+	pds, err := st.PendingDeposits()
+	if err != nil {
+		return nil, errors.Wrap(err, "could not get pending deposits from state")
+	}
+
+	// Return early if there are no pending deposits
+	if len(pds) == 0 {
+		return pdes, nil
+	}
+
+	finalizedEpoch := st.FinalizedCheckpointEpoch()
+	finalizedSlot, err := slots.EpochStart(finalizedEpoch)
+	if err != nil {
+		return nil, errors.Wrap(err, "could not get finalized slot")
+	}
+
+	depBalToConsume, err := st.DepositBalanceToConsume()
+	if err != nil {
+		return nil, errors.Wrap(err, "could not get deposit balance to consume")
+	}
+	availableForProcessing := depBalToConsume + helpers.ActivationBalanceChurnLimit(activeBalance)
+
+	currentEpoch := slots.ToEpoch(st.Slot())
+	processedAmount := uint64(0)
+	nextDepositIndex := uint64(0)
+
+	// state.pending_deposits is guaranteed to be sorted by slot, so we can iterate through it in order
+	// Mostly the same logic as in ProcessPendingDeposits, but it iterates through all pending deposits.
+	for _, pd := range pds {
+		// A pending deposit is processed when its slot is larger than finalized slot.
+		// In this case, moving to the next epoch is needed. This will update and initialize the variables.
+		// Premise: finalizedSlot will always be incremented by 32 slots(= 1 epoch).
+		for pd.Slot > finalizedSlot {
+			currentEpoch += 1
+
+			availableForProcessing = helpers.ActivationBalanceChurnLimit(activeBalance)
+			finalizedSlot += params.BeaconConfig().SlotsPerEpoch
+
+			processedAmount = uint64(0)
+			nextDepositIndex = uint64(0)
+		}
+
+		// Move to next epoch if max pending deposits per epoch is reached.
+		if nextDepositIndex >= params.BeaconConfig().MaxPendingDepositsPerEpoch {
+			currentEpoch += 1
+
+			availableForProcessing = helpers.ActivationBalanceChurnLimit(activeBalance)
```

Comment:
`state.deposit_balance_to_consume` is set to non-zero value only if `is_churn_limit_reached` is true. This line is quite intended, because `is_churn_limit_reached` is calculated later on.